core/src/main/java/io/questdb/griffin/DatabaseSnapshotAgent.java

```diff
@@ -325,4 +211,126 @@ public static void recoverSnapshot(CairoEngine engine) {
+            throw SqlException.position(0).put("Snapshots are not supported on Windows");
+        }
+
+        if (!lock.tryLock()) {
+            throw SqlException.position(0).put("Another snapshot command in progress");
+        }
+        try {
+            if (snapshotReaders.size() > 0) {
+                throw SqlException.position(0).put("Waiting for SNAPSHOT COMPLETE to be called");
+            }
+
+            path.of(configuration.getSnapshotRoot()).concat(configuration.getDbDirectory());
+            int snapshotLen = path.length();
+            // Delete all contents of the snapshot/db dir.
+            if (ff.exists(path.slash$())) {
+                path.trimTo(snapshotLen).$();
+                if (ff.rmdir(path) != 0) {
+                    throw CairoException.instance(ff.errno()).put("Could not remove snapshot dir [dir=").put(path).put(']');
+                }
+            }
+            // Recreate the snapshot/db dir.
+            path.trimTo(snapshotLen).slash$();
+            if (ff.mkdirs(path, configuration.getMkDirMode()) != 0) {
+                throw CairoException.instance(ff.errno()).put("Could not create [dir=").put(path).put(']');
+            }
+
+            try (
+                    TableListRecordCursorFactory factory = new TableListRecordCursorFactory(configuration.getFilesFacade(), configuration.getRoot())
+            ) {
+                final int tableNameIndex = factory.getMetadata().getColumnIndex(TableListRecordCursorFactory.TABLE_NAME_COLUMN);
+                try (RecordCursor cursor = factory.getCursor(executionContext)) {
+                    final Record record = cursor.getRecord();
+                    try (MemoryCMARW mem = Vm.getCMARWInstance()) {
+                        // Copy metadata files for all tables.
+
+                        while (cursor.hasNext()) {
+                            CharSequence tableName = record.getStr(tableNameIndex);
+                            if (
+                                    TableUtils.isValidTableName(tableName, tableName.length())
+                                            && ff.exists(path.of(configuration.getRoot()).concat(tableName).concat(TableUtils.META_FILE_NAME).$())
+                            ) {
+                                path.of(configuration.getSnapshotRoot()).concat(configuration.getDbDirectory());
+                                LOG.info().$("including in the snapshot [table=").$(tableName).$(']').$();
+
+                                TableReader reader = engine.getReaderForStatement(executionContext, tableName, "snapshot");
+                                snapshotReaders.add(reader);
+
+                                path.trimTo(snapshotLen).concat(tableName).slash$();
+                                if (ff.mkdirs(path, configuration.getMkDirMode()) != 0) {
+                                    throw CairoException.instance(ff.errno()).put("Could not create [dir=").put(path).put(']');
```

Comment:
`errno` missing

Reply:
I don't understand, errno is present here `CairoException.instance(ff.errno())`

Reply:
sorry! did not notice!